marker_maze.py

```diff
@@ -0,0 +1,242 @@
+        # Our operations on the frame come here
+        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+        parameters =  aruco.DetectorParameters_create()
+        #print(parameters)
+        '''    detectMarkers(...)
+            detectMarkers(image, dictionary[, corners[, ids[, parameters[, rejectedI
+            mgPoints]]]]) -> corners, ids, rejectedImgPoints
+        '''
+        #lists of ids and the corners beloning to each id
+        corners, ids, rejectedImgPoints = aruco.detectMarkers(gray, self.small_dict, parameters=parameters)
+        if ids != None:
+            #print ("found marker %s" % ids)
+            if len(ids)>1:
+                logger.info( "found %d markers" % len(ids))
+                self.marker_to_track = 0
+                for marker_number in range(0, len(ids)):
+                    if ids[marker_number] == self.turn_number:
+                        self.marker_to_track = marker_number
+                logger.info ("marker I'm looking for, is number %d" % self.marker_to_track)
+            else:
+                self.marker_to_track = 0
+            if ids[self.marker_to_track][0] == self.turn_number:
+                m = self.marker_to_track
+                self.found = True
+                #if found, comptue the centre and move the cursor there
+                found_y = sum([arr[0] for arr in corners[m][0]])  / 4
+                found_x = sum([arr[1] for arr in corners[m][0]])  / 4
+                width = abs(corners[m][0][0][0]-corners[m][0][1][0]+corners[m][0][3][0]-corners[m][0][2][0])/2
+                logger.info('marker width %s' % width)
+                if width > self.TURN_WIDTH[self.turn_number]:
+                    self.turn_number += 1
+                    logger.info('Close to marker making turn %s' % self.turn_number)
+                    if self.turn_number == 5:
+                        logger.info('finished!')
+                        self.drive.move(0,0)
+                        self.finished = True
+                pygame.mouse.set_pos(int(found_x), int(self.CROP_WIDTH-found_y))
+                self.t_error = (self.CROP_WIDTH/2 - found_y) / (self.CROP_WIDTH / 2)
+                turn = self.STEERING_OFFSET + self.TURN_P * self.t_error
+                if self.last_t_error is not 0:
+                    #if there was a real error last time then do some damping
+                    turn -= self.TURN_D *(self.last_t_error - self.t_error)
+                turn = min(max(turn,-self.MAX_TURN_SPEED), self.MAX_TURN_SPEED)
+                #if we're rate limiting the turn, go slow
+                if abs(turn) == self.MAX_TURN_SPEED:
+                    self.drive.move (turn, self.STRAIGHT_SPEED/3)
+                else:
+                    self.drive.move (turn, self.STRAIGHT_SPEED)
+                self.last_t_error = self.t_error
+                #print(camera.exposure_speed)
```

Comment:
💀 ?

Reply:
dead. side note: picamera's exposure_speed is misnamed. it should be exposure time,  as its time in microseconds (large is longer), not fractions of a second like the photography norm (large is shorter). I keep forgetting. I will add an actual note